### PR TITLE
Fix issue 15574 - wrong order of link arguments

### PR DESCRIPTION
**Pull request by rtbo**
This essentially places *.a arguments before the -l and -L switches in
the linker command line.

This allows to link applications that link against a static archive *.a
that depends on -l lib switches.
Such dependency scheme is typical of the following situation:
  - have a D package that contains binding to a C library.
  - this D package is itself built as a static library (libdpackage.a).
  - the C library is built in the same build process as a static library
    (pathtoClib/libclib.a)
  - dmd is invoked in the following way:
```
  $ dmd ... libdpackage.a -L-LpathtoClib -L-lclib
```
Currently dmd reorders arguments and places *.a after -l linker switches
which causes link errors if the -l points to a static library.

This is especially useful with Dub, which can be configured to probe for
the C library, and build it as a static lib if not found system-wide.
In both cases, the C library is provided with "libs" json entry.

So this is not about keeping relative order of linker arguments, it is
about providing an order of linker arguments that is more sound than
the current one.

### CODE REVIEW

**Comment by wilzbach** [comment]
@rtbo have a look at the existing bash scripts in the testsuite for examples on how this can be added to testsuite.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
The documentation failure is unrelated to this PR.

File: test/compilable/issue15574.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+EOF
+
+if [ -z ${CC+x} ]; then CC=cc; fi
+
+$CC -m${MODEL} -c -o ${C_FILE}${OBJ} $C_FILE
+ar rcs ${C_LIB} ${C_FILE}${OBJ}
+
+${DMD} -m${MODEL} -lib -of${D_LIB} ${D_FILE}
+
+${DMD} -m${MODEL} -of${APP} ${APP_FILE} -I${TEST_DIR} ${D_LIB} -L-L${TEST_DIR} -L-lcsquare
```
Comment (RazvanN7) [inline test/compilable/issue15574.sh]: You should also run the executable to make sure that everything is fine.

Comment (rtbo) [reply test/compilable/issue15574.sh]: Done. (and squashed)

File: test/compilable/issue15574.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+${APP}
```
Comment (RazvanN7) [inline test/compilable/issue15574.sh]: It would be best if you could check the application actually runs correctly i.e. it finished with exit code 0

Comment (rtbo) [reply test/compilable/issue15574.sh]: Doesn't the test fail if a command returns non-zero?

Comment (RazvanN7) [reply test/compilable/issue15574.sh]: you are correct.

**Comment by RazvanN7** [comment]
@rtbo I added the `merge` label on this PR. Thank you for your contribution

**Comment by CyberShadow** [comment]
@rtbo This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=19243

**Pull request merged**